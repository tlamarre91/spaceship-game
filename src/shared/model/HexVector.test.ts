import {
  HexVector,
  HEX_DIRECTIONS
} from "./HexVector";

const box = {
  xMin: -1000,
  xMax: 1000,
  yMin: -1000,
  yMax: 1000,
};

describe("HexVector (basic)", () => {
  it("validates cubic coordinates", () => {
    const f = () => {
      const v = HexVector.fromCubicCoordinates(4, 2, 2341);
    };

    expect(f).toThrowError("invalid cubic coordinates: (4, 2, 2341)");
  });

  it("HexVector.EPSILON is considered equal to HexVector.ZERO", () => {
    const v = HexVector.EPSILON_CUBIC;
    const w = HexVector.ZERO;
    expect(v.equals(w)).toBeTruthy();
    expect(w.equals(v)).toBeTruthy();
  });

  it("HexVectors with different coordinates are not equal", () => {
    const v = new HexVector("a", 0.0003, 0.0005);
    const w = new HexVector("a", 0.0005, 0.0003);
    expect(v.equals(w)).toBeFalsy();
  });

  it("different HexVectors are not equal", () => {
    HEX_DIRECTIONS.forEach((dir) => {
      const v = HexVector.direction(dir);
      expect(v.equals(HexVector.ZERO)).toBeFalsy();
    });
  });

  it("HexVector equals itself", () => {
    const v = new HexVector("c", 30, 40, -70);
    expect(v.equals(v)).toBeTruthy();
  });

  it("HexVector equals its equivalent in the other system", () => {
    const v = new HexVector("c", 30, 40, -70);
    const w = new HexVector("a", 30, 40);
    expect(v.equals(w)).toBeTruthy();
    expect(w.equals(v)).toBeTruthy();
  });

  it("adds axial", () => {
    const v = HexVector.fromAxialCoordinates(5, 2);
    const w = HexVector.fromAxialCoordinates(43, 2);
    expect(v.plus(w).equals(w.plus(v))).toBeTruthy();
    expect(v.plus(w)).toMatchInlineSnapshot(`
      HexVector {
        "system": "a",
        "x": 48,
        "y": 4,
        "z": null,
      }
    `);
  });

  it("adds cubic", () => {
    const v = HexVector.fromCubicCoordinates(5, -2, -3);
    const w = HexVector.fromCubicCoordinates(233, 7, -240);
    expect(v.plus(w).equals(w.plus(v))).toBeTruthy();
    expect(v.plus(w)).toMatchInlineSnapshot(`
      HexVector {
        "system": "c",
        "x": 238,
        "y": 5,
        "z": -243,
      }
    `);
  });
});

describe("HexVector (randomized)", () => {
  it("equals itself", () => {
    const v = HexVector.random(box);
    expect(v.equals(v)).toBeTruthy();
  });

  it("adds axial", () => {
    const v = HexVector.random(box);
    const w = HexVector.random(box);

    const { x: x0, y: y0 } = v;
    const { x: x1, y: y1 } = w;

    const sum = v.plus(w);
    expect(sum.x).toBe(x0 + x1);
    expect(sum.y).toBe(y0 + y1);
  });

  it("adds cubic", () => {
    const v = HexVector.random(box).toCubic();
    const w = HexVector.random(box).toCubic();

    const { x: x0, y: y0, z: z0 } = v;
    const { x: x1, y: y1, z: z1 } = w;

    const sum = v.plus(w);
    expect(sum.x).toBe(x0 + x1);
    expect(sum.y).toBe(y0 + y1);
    expect(sum.z).toBe(z0 + z1);
  });
});

describe("HexVector (methods)", () => {
  it.skip("pathTo() generates a grid path between two vectors", () => {
  });
});
