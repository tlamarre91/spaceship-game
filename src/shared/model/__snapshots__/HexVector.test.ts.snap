// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HexVector (methods) pathTo() generates a grid path between two vectors 1`] = `
Array [
  HexVector {
    "system": "c",
    "x": 5,
    "y": -10,
    "z": 5,
  },
  HexVector {
    "system": "c",
    "x": 4,
    "y": -9,
    "z": 5,
  },
  HexVector {
    "system": "c",
    "x": 3,
    "y": -8,
    "z": 5,
  },
  HexVector {
    "system": "c",
    "x": 3,
    "y": -7,
    "z": 4,
  },
  HexVector {
    "system": "c",
    "x": 2,
    "y": -6,
    "z": 4,
  },
  HexVector {
    "system": "c",
    "x": 1,
    "y": -5,
    "z": 4,
  },
  HexVector {
    "system": "c",
    "x": 0,
    "y": -4,
    "z": 4,
  },
  HexVector {
    "system": "c",
    "x": -1,
    "y": -3,
    "z": 4,
  },
  HexVector {
    "system": "c",
    "x": -2,
    "y": -2,
    "z": 4,
  },
  HexVector {
    "system": "c",
    "x": -2,
    "y": -1,
    "z": 3,
  },
  HexVector {
    "system": "c",
    "x": -3,
    "y": 0,
    "z": 3,
  },
  HexVector {
    "system": "c",
    "x": -4,
    "y": 1,
    "z": 3,
  },
]
`;
